# Exploring the Dataset:

In [559]:
# import statements:

import pandas as pd

# Exploring the Dataset:

In [560]:
# reading the dataset with pandas and turning it into a pandas DF:

df = pd.read_csv('../data/googleplaystore.csv')

Having a general look at the dataset:


In [561]:
# looking at the numer of rows and columns in the dataset:

df.shape

(10841, 13)

In [562]:
# looking at the dataframe'c columns and the dtypes:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


we have some null values in the dataframe!

In [563]:
# looking at the first 5 rows of the dataframe to see how the columns are filled:

df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Cleaning up the dataframe:

#### Duplicated Rows:

In [564]:
# checking the number of duplicated rows:

df.duplicated().sum()

483

we have 483 duplicated rows, so we need to drop them entirely.

In [565]:
#dropping the duplicated rows and storing the resulting entities in a new dataframe:

clean_df = df.drop_duplicates()

In [566]:
# making sure that we got rid of the duplicated rows:

clean_df.duplicated().sum()

0

let us check for duplicates in the app names and see if they are different entities:

In [567]:
clean_df['App'].duplicated().sum()

698

We have 698 duplicated Apps, let's check them out!

In [568]:
clean_df[clean_df['App'].duplicated() == True]['App'].value_counts()

App
ROBLOX                            8
8 Ball Pool                       6
Bubble Shooter                    5
Zombie Catchers                   5
Helix Jump                        5
                                 ..
Google Docs                       1
Microsoft Excel                   1
Smart Launcher 5                  1
Calculator                        1
Slickdeals: Coupons & Shopping    1
Name: count, Length: 523, dtype: int64

In [569]:
clean_df[clean_df['App']== 'ROBLOX']

App Category  Rating  Reviews Size      Installs  Type Price  \
1653  ROBLOX     GAME     4.5  4447388  67M  100,000,000+  Free     0   
1701  ROBLOX     GAME     4.5  4447346  67M  100,000,000+  Free     0   
1748  ROBLOX     GAME     4.5  4448791  67M  100,000,000+  Free     0   
1841  ROBLOX     GAME     4.5  4449882  67M  100,000,000+  Free     0   
1870  ROBLOX     GAME     4.5  4449910  67M  100,000,000+  Free     0   
2016  ROBLOX   FAMILY     4.5  4449910  67M  100,000,000+  Free     0   
2088  ROBLOX   FAMILY     4.5  4450855  67M  100,000,000+  Free     0   
2206  ROBLOX   FAMILY     4.5  4450890  67M  100,000,000+  Free     0   
4527  ROBLOX   FAMILY     4.5  4443407  67M  100,000,000+  Free     0   

     Content Rating                        Genres   Last Updated  \
1653   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1701   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1748   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1841   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1870   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2016   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2088   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2206   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
4527   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   

       Current Ver Android Ver  
1653  2.347.225742  4.1 and up  
1701  2.347.225742  4.1 and up  
1748  2.347.225742  4.1 and up  
1841  2.347.225742  4.1 and up  
1870  2.347.225742  4.1 and up  
2016  2.347.225742  4.1 and up  
2088  2.347.225742  4.1 and up  
2206  2.347.225742  4.1 and up  
4527  2.347.225742  4.1 and up

In [570]:
clean_df[clean_df['App']== '8 Ball Pool']

App Category  Rating   Reviews Size      Installs  Type Price  \
1675  8 Ball Pool     GAME     4.5  14198297  52M  100,000,000+  Free     0   
1703  8 Ball Pool     GAME     4.5  14198602  52M  100,000,000+  Free     0   
1755  8 Ball Pool     GAME     4.5  14200344  52M  100,000,000+  Free     0   
1844  8 Ball Pool     GAME     4.5  14200550  52M  100,000,000+  Free     0   
1871  8 Ball Pool     GAME     4.5  14201891  52M  100,000,000+  Free     0   
1970  8 Ball Pool     GAME     4.5  14201604  52M  100,000,000+  Free     0   
3953  8 Ball Pool   SPORTS     4.5  14184910  52M  100,000,000+  Free     0   

     Content Rating  Genres   Last Updated Current Ver   Android Ver  
1675       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1703       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1755       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1844       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1871       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1970       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
3953       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up

In [571]:
clean_df[clean_df['App']== 'Google Docs']

App      Category  Rating Reviews                Size  \
3477  Google Docs  PRODUCTIVITY     4.3  815981  Varies with device   
3528  Google Docs  PRODUCTIVITY     4.3  815974  Varies with device   

          Installs  Type Price Content Rating        Genres    Last Updated  \
3477  100,000,000+  Free     0       Everyone  Productivity  August 6, 2018   
3528  100,000,000+  Free     0       Everyone  Productivity  August 6, 2018   

             Current Ver         Android Ver  
3477  Varies with device  Varies with device  
3528  Varies with device  Varies with device

We can see that the apps differ mostly by the number of reviews and categories. I will drop the duplicates according to the number of reviews. 

We will be keeping the entry with the highest number of reviews!

To do so we have to go through a number of steps:

In [572]:
# sorting the df by app name and number of reviews:

clean_df = clean_df.sort_values(by=['App', 'Reviews'], ascending=[True, False])

In [573]:
# dropping duplicates on the App name level and just keeping the first entry which will be the one with the highest number of reviews:

clean_df = clean_df.drop_duplicates(subset=['App'], keep='first')

In [574]:
# making sure this worked:

clean_df['App'].duplicated().sum()

0

now that we are done with duplicates, let us see what we can do with the null values:

### Null Values:

In [575]:
# getting the number of null values in the dataframe:

clean_df.isna().sum()

App                  0
Category             0
Rating            1463
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

let us look first at the lesser null values:

In [576]:
# getting the rows with null values in the Type column:

clean_df[clean_df['Type'].isna()==True]

App Category  Rating Reviews                Size  \
9148  Command & Conquer: Rivals   FAMILY     NaN       0  Varies with device   

     Installs Type Price Content Rating    Genres   Last Updated  \
9148        0  NaN     0   Everyone 10+  Strategy  June 28, 2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

In [577]:
# replacing the null value with the app type!

clean_df.loc[clean_df['App'] == 'Command & Conquer: Rivals', 'Type'] = 'Free'


In [578]:
# making sure the previous step worked

clean_df['Type'].isna().sum()

0

In [579]:
# getting the rows with null values in the Content Rating column:

clean_df[clean_df['Content Rating'].isna()==True]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [580]:
# replacing the null value:

clean_df.loc[clean_df['App'] == 'Life Made WI-Fi Touchscreen Photo Frame', 'Content Rating'] = 'Everyone'

In [581]:
# checking if it worked:

clean_df['Content Rating'].isna().sum()

0

In [582]:
# getting the rows with null values in the Curr Ver. column:

clean_df[clean_df['Current Ver'].isna()==True]

App             Category  Rating Reviews  \
6803                         BT Master               FAMILY     NaN       0   
7407                   Calculate My IQ               FAMILY     NaN      44   
7333                       Dots puzzle               FAMILY     4.0     179   
10342                   La Fe de Jesus  BOOKS_AND_REFERENCE     NaN       8   
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN     3.2      55   
1553              Market Update Helper   LIBRARIES_AND_DEMO     4.1   20145   
7730                            UFO-CQ                TOOLS     NaN       1   
6322            Virtual DJ Sound Mixer                TOOLS     4.2    4010   

       Size    Installs  Type  Price Content Rating             Genres  \
6803   222k        100+  Free      0       Everyone          Education   
7407   7.2M     10,000+  Free      0       Everyone      Entertainment   
7333    14M     50,000+  Paid  $0.99       Everyone             Puzzle   
10342  658k      1,000+  Free      0       Everyone  Books & Reference   
15     2.7M      5,000+  Free      0       Everyone       Art & Design   
1553    11k  1,000,000+  Free      0       Everyone   Libraries & Demo   
7730   237k         10+  Paid  $0.99       Everyone              Tools   
6322   8.7M    500,000+  Free      0       Everyone              Tools   

            Last Updated Current Ver Android Ver  
6803    November 6, 2016         NaN  1.6 and up  
7407       April 3, 2017         NaN  2.3 and up  
7333      April 18, 2018         NaN  4.0 and up  
10342   January 31, 2017         NaN  3.0 and up  
15          June 6, 2018         NaN  4.2 and up  
1553   February 12, 2013         NaN  1.5 and up  
7730        July 4, 2016         NaN  2.0 and up  
6322        May 10, 2017         NaN  4.0 and up

This will be hard to get and fill so we will drop the null values!

Also for Android Ver!

In [583]:
# dropping the null values in the two specified columns:

clean_df = clean_df.dropna(subset=['Current Ver', 'Android Ver'])

In [584]:
# checking if it worked:

clean_df.isna().sum()

App                  0
Category             0
Rating            1459
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
dtype: int64

Now to see what to do with the rating column as the null values are 15% of the dataframe!

In [585]:
clean_df[clean_df['Rating'].isna()==True]

App       Category  \
8884                "i DT" Fútbol. Todos Somos Técnicos.         SPORTS   
2483                                       14thStreetVet        MEDICAL   
7382                           1st Fed CI Mobile Banking        FINANCE   
6518                         2017 BN SM Sales Conference       BUSINESS   
9660                                      23rd QM BDE EO       BUSINESS   
...                                                  ...            ...   
321    【Ranobbe complete free】 Novelba - Free app tha...         COMICS   
9466                     パーリーゲイツ公式通販｜EJ STYLE（イージェイスタイル）       SHOPPING   
5513                                         中国語 AQリスニング         FAMILY   
10109                                            哈哈姆特不EY  COMMUNICATION   
6334                             뽕티비 - 개인방송, 인터넷방송, BJ방송  VIDEO_PLAYERS   

       Rating Reviews  Size  Installs  Type Price Content Rating  \
8884      NaN      27  3.6M      500+  Free     0       Everyone   
2483      NaN       0   29M        5+  Free     0       Everyone   
7382      NaN       0   31M      500+  Free     0       Everyone   
6518      NaN       2   10M      100+  Free     0           Teen   
9660      NaN       0  6.3M       10+  Free     0       Everyone   
...       ...     ...   ...       ...   ...   ...            ...   
321       NaN    1330   22M   50,000+  Free     0       Everyone   
9466      NaN       1  9.3M      100+  Free     0       Everyone   
5513      NaN      21   17M    5,000+  Free     0       Everyone   
10109     NaN     239   18M   10,000+  Free     0       Everyone   
6334      NaN     414   59M  100,000+  Free     0     Mature 17+   

                        Genres        Last Updated  Current Ver   Android Ver  
8884                    Sports     October 7, 2017         0.22    4.1 and up  
2483                   Medical       July 16, 2018  300000.1.11  4.0.3 and up  
7382                   Finance  September 29, 2016   6.0.0.1775    4.1 and up  
6518                  Business  September 28, 2017    v2.7.11.6    4.1 and up  
9660                  Business  September 27, 2017        1.0.0    4.1 and up  
...                        ...                 ...          ...           ...  
321                     Comics        July 3, 2018        6.1.1    4.2 and up  
9466                  Shopping   November 16, 2017          3.0  4.0.3 and up  
5513                 Education       June 22, 2016        2.4.0    4.0 and up  
10109            Communication       July 31, 2018        1.3.4    4.4 and up  
6334   Video Players & Editors       July 18, 2018        4.0.7  4.0.3 and up  

[1459 rows x 13 columns]

Sadly, we have to also drop the nulls in this column.

In [586]:
# dropping the nulls in the Rating column:

clean_df = clean_df.dropna(subset=['Rating'])

In [587]:
# checking if we have any null values left:

clean_df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

No more null values!

Now we can move to the next step which is fixing the columns and their data types!

### Fixing Up Columns + Dtypes:

we can fill this by external research!

With research, we found out that this game is free!

In [588]:
# having a general look at the data again:

clean_df.head()

App            Category  Rating Reviews  Size  \
8532  +Download 4 Instagram Twitter              SOCIAL     4.5   40467   22M   
324      - Free Comics - Comic Apps              COMICS     3.5     115  9.1M   
4541                             .R               TOOLS     4.5     259  203k   
4636                         /u/app       COMMUNICATION     4.7     573   53M   
5940                         058.ba  NEWS_AND_MAGAZINES     4.4      27   14M   

        Installs  Type Price Content Rating            Genres  \
8532  1,000,000+  Free     0       Everyone            Social   
324      10,000+  Free     0     Mature 17+            Comics   
4541     10,000+  Free     0       Everyone             Tools   
4636     10,000+  Free     0     Mature 17+     Communication   
5940        100+  Free     0       Everyone  News & Magazines   

            Last Updated Current Ver Android Ver  
8532      August 2, 2018        5.03  4.1 and up  
324        July 13, 2018      5.0.12  5.0 and up  
4541  September 16, 2014      1.1.06  1.5 and up  
4636        July 3, 2018       4.2.4  4.1 and up  
5940        July 6, 2018         1.0  4.2 and up

#### Category Column:

In [589]:
# counting the number of unique categories:

clean_df['Category'].nunique()

33

In [590]:
# getting the names of the categories:

clean_df['Category'].value_counts()

Category
FAMILY                 1650
GAME                    896
TOOLS                   718
FINANCE                 302
LIFESTYLE               301
PRODUCTIVITY            301
PERSONALIZATION         296
MEDICAL                 290
PHOTOGRAPHY             263
BUSINESS                263
SPORTS                  260
COMMUNICATION           256
HEALTH_AND_FITNESS      244
NEWS_AND_MAGAZINES      204
SOCIAL                  203
TRAVEL_AND_LOCAL        187
SHOPPING                180
BOOKS_AND_REFERENCE     169
VIDEO_PLAYERS           149
DATING                  134
MAPS_AND_NAVIGATION     118
EDUCATION               107
FOOD_AND_DRINK           94
ENTERTAINMENT            87
AUTO_AND_VEHICLES        73
WEATHER                  72
LIBRARIES_AND_DEMO       63
HOUSE_AND_HOME           61
ART_AND_DESIGN           58
COMICS                   54
PARENTING                50
EVENTS                   45
BEAUTY                   42
Name: count, dtype: int64

In [591]:
clean_df['Category'].unique()

array(['SOCIAL', 'COMICS', 'TOOLS', 'COMMUNICATION', 'NEWS_AND_MAGAZINES',
       'SPORTS', 'HEALTH_AND_FITNESS', 'FAMILY', 'BOOKS_AND_REFERENCE',
       'LIFESTYLE', 'BUSINESS', 'SHOPPING', 'MEDICAL', 'GAME',
       'PERSONALIZATION', 'PHOTOGRAPHY', 'TRAVEL_AND_LOCAL', 'DATING',
       'PRODUCTIVITY', 'ART_AND_DESIGN', 'FOOD_AND_DRINK',
       'VIDEO_PLAYERS', 'HOUSE_AND_HOME', 'MAPS_AND_NAVIGATION',
       'FINANCE', 'EDUCATION', 'AUTO_AND_VEHICLES', 'WEATHER', 'BEAUTY',
       'ENTERTAINMENT', 'EVENTS', 'LIBRARIES_AND_DEMO', 'PARENTING'],
      dtype=object)

we can see in some categories that the space is replaced with a '_', let us fix that:

In [592]:
# replacing '_' with ' ':

clean_df['Category'] = clean_df['Category'].str.replace('_', ' ')

In [593]:
# checking if it worked:

clean_df['Category'].value_counts()

Category
FAMILY                 1650
GAME                    896
TOOLS                   718
FINANCE                 302
LIFESTYLE               301
PRODUCTIVITY            301
PERSONALIZATION         296
MEDICAL                 290
PHOTOGRAPHY             263
BUSINESS                263
SPORTS                  260
COMMUNICATION           256
HEALTH AND FITNESS      244
NEWS AND MAGAZINES      204
SOCIAL                  203
TRAVEL AND LOCAL        187
SHOPPING                180
BOOKS AND REFERENCE     169
VIDEO PLAYERS           149
DATING                  134
MAPS AND NAVIGATION     118
EDUCATION               107
FOOD AND DRINK           94
ENTERTAINMENT            87
AUTO AND VEHICLES        73
WEATHER                  72
LIBRARIES AND DEMO       63
HOUSE AND HOME           61
ART AND DESIGN           58
COMICS                   54
PARENTING                50
EVENTS                   45
BEAUTY                   42
Name: count, dtype: int64

#### Size Column:

In [594]:
clean_df['Size'].value_counts()

Size
Varies with device    1169
14M                    148
12M                    146
13M                    143
11M                    143
                      ... 
89k                      1
914k                     1
784k                     1
951k                     1
655k                     1
Name: count, Length: 413, dtype: int64

In [595]:
clean_df['Size'] = clean_df['Size'].str.replace('Varies with device', '0')
clean_df['Size'] = clean_df['Size'].str.replace('M', '000')
clean_df['Size'] = clean_df['Size'].str.replace('.', '')
clean_df['Size'] = clean_df['Size'].str.replace('k', '')

In [596]:
clean_df['Size'] = clean_df['Size'].astype(int)

#### Reviews Column:

we just have to change the data type of the reviews column!

In [597]:
# changing the data type of the reviews column into an integer:

clean_df['Reviews'] = clean_df['Reviews'].astype(int)

#### Price Column:

We fix it up first removing the $ sign ang then tunring the price into a numeric

In [598]:
# removing the $ sign

clean_df['Price'] = clean_df['Price'].str.replace('$', '')

In [599]:
# changing the price column dtype to float:

clean_df['Price'] = clean_df['Price'].astype(float)

#### Last Updated Column:

In [600]:
# changing the data type of the last updated column to datetime:

clean_df['Last Updated'] = pd.to_datetime(clean_df['Last Updated'], format='mixed')

#### Removing Some Columns:

In [601]:
clean_df = clean_df.drop(['Android Ver', 'Current Ver'], axis=1)

#### Installs Column:

we will be turning the installs column into the label column

In [602]:
clean_df['Installs'].value_counts()

Installs
1,000,000+        1415
100,000+          1095
10,000+            985
10,000,000+        934
1,000+             696
5,000,000+         608
500,000+           503
50,000+            456
5,000+             424
100+               303
50,000,000+        203
500+               199
100,000,000+       188
10+                 69
50+                 56
500,000,000+        24
1,000,000,000+      20
5+                   9
1+                   3
Name: count, dtype: int64

In [603]:
clean_df['Installs'] = clean_df['Installs'].str.replace('+', '')
clean_df['Installs'] = clean_df['Installs'].str.replace(',', '')

In [604]:
clean_df['Installs'] = clean_df['Installs'].astype(int)

In [605]:
clean_df['Installs'].loc[clean_df['Installs']<100000] = 0

C:\Users\allab\AppData\Local\Temp\ipykernel_10604\1772492956.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  clean_df['Installs'].loc[clean_df['Installs']<100000] = 0
C:\Users\allab\AppData\Local\Temp\ipykernel_10604\1772492956.py:1: Set

In [606]:
clean_df['Installs'].loc[clean_df['Installs']>=100000] = 1

C:\Users\allab\AppData\Local\Temp\ipykernel_10604\4079046151.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  clean_df['Installs'].loc[clean_df['Installs']>=100000] = 1
C:\Users\allab\AppData\Local\Temp\ipykernel_10604\4079046151.py:1: Se

In [607]:
clean_df['Installs'].value_counts()

Installs
1    4990
0    3200
Name: count, dtype: int64

In [608]:
clean_df.rename(columns={'Installs': 'Label'}, inplace=True)

#### Type Column:

In [609]:
clean_df['Type'] = clean_df['Type'].map({'Free': 0, 'Paid': 1})

In [610]:
clean_df.rename(columns={'Type': 'Paid'}, inplace=True)

#### Genres Column:

In [611]:
splited = clean_df

In [612]:
# spliting the concatenated genres:

splited[['genre_1', 'genre_2']] = clean_df['Genres'].str.split(';', n=1, expand=True)

In [613]:
# seeing how many unique genres we have:

split_genres = clean_df['Genres'].str.split(';')

s = set()
for i in split_genres:
    for genre in i:
        s.add(genre)
print(len(s))

53


In [614]:
genres_1 = pd.get_dummies(splited['genre_1'])
genres_2 = pd.get_dummies(splited['genre_2'])

In [615]:
genres_1


Action  Adventure  Arcade  Art & Design  Auto & Vehicles  Beauty  Board  \
8532   False      False   False         False            False   False  False   
324    False      False   False         False            False   False  False   
4541   False      False   False         False            False   False  False   
4636   False      False   False         False            False   False  False   
5940   False      False   False         False            False   False  False   
...      ...        ...     ...           ...              ...     ...    ...   
6420   False      False   False         False            False   False  False   
4362   False      False   False         False            False   False  False   
2575   False      False   False         False            False   False  False   
7559   False      False   False         False            False   False  False   
882    False      False   False         False            False   False  False   

      Books & Reference  Business   Card  ...  Simulation  Social  Sports  \
8532              False     False  False  ...       False    True   False   
324               False     False  False  ...       False   False   False   
4541              False     False  False  ...       False   False   False   
4636              False     False  False  ...       False   False   False   
5940              False     False  False  ...       False   False   False   
...                 ...       ...    ...  ...         ...     ...     ...   
6420              False     False  False  ...       False   False   False   
4362              False     False  False  ...       False   False   False   
2575              False     False  False  ...       False    True   False   
7559              False     False  False  ...       False   False   False   
882               False     False  False  ...       False   False   False   

      Strategy  Tools  Travel & Local  Trivia  Video Players & Editors  \
8532     False  False           False   False                    False   
324      False  False           False   False                    False   
4541     False   True           False   False                    False   
4636     False  False           False   False                    False   
5940     False  False           False   False                    False   
...        ...    ...             ...     ...                      ...   
6420     False  False           False   False                    False   
4362     False  False           False   False                    False   
2575     False  False           False   False                    False   
7559     False   True           False   False                    False   
882      False  False           False   False                    False   

      Weather   Word  
8532    False  False  
324     False  False  
4541    False  False  
4636    False  False  
5940    False  False  
...       ...    ...  
6420    False  False  
4362    False  False  
2575    False  False  
7559    False  False  
882     False  False  

[8190 rows x 48 columns]

In [616]:
genres_2

Action & Adventure  Brain Games  Creativity  Education  Music & Video  \
8532               False        False       False      False          False   
324                False        False       False      False          False   
4541               False        False       False      False          False   
4636               False        False       False      False          False   
5940               False        False       False      False          False   
...                  ...          ...         ...        ...            ...   
6420               False        False       False      False          False   
4362               False        False       False      False          False   
2575               False        False       False      False          False   
7559               False        False       False      False          False   
882                False        False       False      False          False   

      Pretend Play  
8532         False  
324          False  
4541         False  
4636         False  
5940         False  
...            ...  
6420         False  
4362         False  
2575         False  
7559         False  
882          False  

[8190 rows x 6 columns]

In [617]:
splited = splited.drop(['Genres', 'genre_1', 'genre_2'], axis=1)

In [618]:
merged_splits = pd.concat([splited,genres_1, genres_2], axis=1)

In [619]:
merged_splits

App            Category  \
8532                     +Download 4 Instagram Twitter              SOCIAL   
324                         - Free Comics - Comic Apps              COMICS   
4541                                                .R               TOOLS   
4636                                            /u/app       COMMUNICATION   
5940                                            058.ba  NEWS AND MAGAZINES   
...                                                ...                 ...   
6420                                       감성학원 BL 첫사랑              COMICS   
4362                                        💎 I'm rich           LIFESTYLE   
2575     💘 WhatsLov: Smileys of love, stickers and GIF              SOCIAL   
7559  📏 Smart Ruler ↔️ cm/inch measuring for homework!               TOOLS   
882   🔥 Football Wallpapers 4K | Full HD Backgrounds 😍       ENTERTAINMENT   

      Rating  Reviews   Size  Label  Paid   Price Content Rating Last Updated  \
8532     4.5    40467  22000      1     0    0.00       Everyone   2018-08-02   
324      3.5      115  91000      0     0    0.00     Mature 17+   2018-07-13   
4541     4.5      259    203      0     0    0.00       Everyone   2014-09-16   
4636     4.7      573  53000      0     0    0.00     Mature 17+   2018-07-03   
5940     4.4       27  14000      0     0    0.00       Everyone   2018-07-06   
...      ...      ...    ...    ...   ...     ...            ...          ...   
6420     4.4      190  34000      0     0    0.00       Everyone   2017-11-17   
4362     3.8      718  26000      0     1  399.99       Everyone   2018-03-11   
2575     4.6    22098  18000      1     0    0.00       Everyone   2018-07-24   
7559     4.0       19  32000      0     0    0.00       Everyone   2017-10-21   
882      4.7    11661  40000      1     0    0.00       Everyone   2018-07-14   

      ...  Trivia  Video Players & Editors  Weather   Word  \
8532  ...   False                    False    False  False   
324   ...   False                    False    False  False   
4541  ...   False                    False    False  False   
4636  ...   False                    False    False  False   
5940  ...   False                    False    False  False   
...   ...     ...                      ...      ...    ...   
6420  ...   False                    False    False  False   
4362  ...   False                    False    False  False   
2575  ...   False                    False    False  False   
7559  ...   False                    False    False  False   
882   ...   False                    False    False  False   

      Action & Adventure  Brain Games  Creativity  Education  Music & Video  \
8532               False        False       False      False          False   
324                False        False       False      False          False   
4541               False        False       False      False          False   
4636               False        False       False      False          False   
5940               False        False       False      False          False   
...                  ...          ...         ...        ...            ...   
6420               False        False       False      False          False   
4362               False        False       False      False          False   
2575               False        False       False      False          False   
7559               False        False       False      False          False   
882                False        False       False      False          False   

      Pretend Play  
8532         False  
324          False  
4541         False  
4636         False  
5940         False  
...            ...  
6420         False  
4362         False  
2575         False  
7559         False  
882          False  

[8190 rows x 64 columns]

#### Turning the Categories into Columns:

In [620]:
categories = pd.get_dummies(clean_df['Category'])

In [621]:
merged_splits = pd.concat([merged_splits, categories], axis=1)

In [622]:
merged_splits = merged_splits.drop('Category', axis=1)

In [623]:
merged_splits

App  Rating  Reviews  \
8532                     +Download 4 Instagram Twitter     4.5    40467   
324                         - Free Comics - Comic Apps     3.5      115   
4541                                                .R     4.5      259   
4636                                            /u/app     4.7      573   
5940                                            058.ba     4.4       27   
...                                                ...     ...      ...   
6420                                       감성학원 BL 첫사랑     4.4      190   
4362                                        💎 I'm rich     3.8      718   
2575     💘 WhatsLov: Smileys of love, stickers and GIF     4.6    22098   
7559  📏 Smart Ruler ↔️ cm/inch measuring for homework!     4.0       19   
882   🔥 Football Wallpapers 4K | Full HD Backgrounds 😍     4.7    11661   

       Size  Label  Paid   Price Content Rating Last Updated  Action  ...  \
8532  22000      1     0    0.00       Everyone   2018-08-02   False  ...   
324   91000      0     0    0.00     Mature 17+   2018-07-13   False  ...   
4541    203      0     0    0.00       Everyone   2014-09-16   False  ...   
4636  53000      0     0    0.00     Mature 17+   2018-07-03   False  ...   
5940  14000      0     0    0.00       Everyone   2018-07-06   False  ...   
...     ...    ...   ...     ...            ...          ...     ...  ...   
6420  34000      0     0    0.00       Everyone   2017-11-17   False  ...   
4362  26000      0     1  399.99       Everyone   2018-03-11   False  ...   
2575  18000      1     0    0.00       Everyone   2018-07-24   False  ...   
7559  32000      0     0    0.00       Everyone   2017-10-21   False  ...   
882   40000      1     0    0.00       Everyone   2018-07-14   False  ...   

      PERSONALIZATION  PHOTOGRAPHY  PRODUCTIVITY  SHOPPING  SOCIAL  SPORTS  \
8532            False        False         False     False    True   False   
324             False        False         False     False   False   False   
4541            False        False         False     False   False   False   
4636            False        False         False     False   False   False   
5940            False        False         False     False   False   False   
...               ...          ...           ...       ...     ...     ...   
6420            False        False         False     False   False   False   
4362            False        False         False     False   False   False   
2575            False        False         False     False    True   False   
7559            False        False         False     False   False   False   
882             False        False         False     False   False   False   

      TOOLS  TRAVEL AND LOCAL  VIDEO PLAYERS  WEATHER  
8532  False             False          False    False  
324   False             False          False    False  
4541   True             False          False    False  
4636  False             False          False    False  
5940  False             False          False    False  
...     ...               ...            ...      ...  
6420  False             False          False    False  
4362  False             False          False    False  
2575  False             False          False    False  
7559   True             False          False    False  
882   False             False          False    False  

[8190 rows x 96 columns]

#### Content Rating:

In [624]:
content_rating = pd.get_dummies(clean_df['Content Rating'])

In [625]:
merged_splits = merged_splits.drop('Content Rating', axis=1)

In [626]:
merged_splits = pd.concat([merged_splits, content_rating], axis=1)

In [627]:
merged_splits

App  Rating  Reviews  \
8532                     +Download 4 Instagram Twitter     4.5    40467   
324                         - Free Comics - Comic Apps     3.5      115   
4541                                                .R     4.5      259   
4636                                            /u/app     4.7      573   
5940                                            058.ba     4.4       27   
...                                                ...     ...      ...   
6420                                       감성학원 BL 첫사랑     4.4      190   
4362                                        💎 I'm rich     3.8      718   
2575     💘 WhatsLov: Smileys of love, stickers and GIF     4.6    22098   
7559  📏 Smart Ruler ↔️ cm/inch measuring for homework!     4.0       19   
882   🔥 Football Wallpapers 4K | Full HD Backgrounds 😍     4.7    11661   

       Size  Label  Paid   Price Last Updated  Action  Adventure  ...  TOOLS  \
8532  22000      1     0    0.00   2018-08-02   False      False  ...  False   
324   91000      0     0    0.00   2018-07-13   False      False  ...  False   
4541    203      0     0    0.00   2014-09-16   False      False  ...   True   
4636  53000      0     0    0.00   2018-07-03   False      False  ...  False   
5940  14000      0     0    0.00   2018-07-06   False      False  ...  False   
...     ...    ...   ...     ...          ...     ...        ...  ...    ...   
6420  34000      0     0    0.00   2017-11-17   False      False  ...  False   
4362  26000      0     1  399.99   2018-03-11   False      False  ...  False   
2575  18000      1     0    0.00   2018-07-24   False      False  ...  False   
7559  32000      0     0    0.00   2017-10-21   False      False  ...   True   
882   40000      1     0    0.00   2018-07-14   False      False  ...  False   

      TRAVEL AND LOCAL  VIDEO PLAYERS  WEATHER  Adults only 18+  Everyone  \
8532             False          False    False            False      True   
324              False          False    False            False     False   
4541             False          False    False            False      True   
4636             False          False    False            False     False   
5940             False          False    False            False      True   
...                ...            ...      ...              ...       ...   
6420             False          False    False            False      True   
4362             False          False    False            False      True   
2575             False          False    False            False      True   
7559             False          False    False            False      True   
882              False          False    False            False      True   

      Everyone 10+  Mature 17+   Teen  Unrated  
8532         False       False  False    False  
324          False        True  False    False  
4541         False       False  False    False  
4636         False        True  False    False  
5940         False       False  False    False  
...            ...         ...    ...      ...  
6420         False       False  False    False  
4362         False       False  False    False  
2575         False       False  False    False  
7559         False       False  False    False  
882          False       False  False    False  

[8190 rows x 101 columns]

------------------------------------------------------------------------------------------------------------------------------------------


# Data sets to experiment on:

#### All one hot encoding:

In [628]:
merged_splits.head()

App  Rating  Reviews   Size  Label  Paid  \
8532  +Download 4 Instagram Twitter     4.5    40467  22000      1     0   
324      - Free Comics - Comic Apps     3.5      115  91000      0     0   
4541                             .R     4.5      259    203      0     0   
4636                         /u/app     4.7      573  53000      0     0   
5940                         058.ba     4.4       27  14000      0     0   

      Price Last Updated  Action  Adventure  ...  TOOLS  TRAVEL AND LOCAL  \
8532    0.0   2018-08-02   False      False  ...  False             False   
324     0.0   2018-07-13   False      False  ...  False             False   
4541    0.0   2014-09-16   False      False  ...   True             False   
4636    0.0   2018-07-03   False      False  ...  False             False   
5940    0.0   2018-07-06   False      False  ...  False             False   

      VIDEO PLAYERS  WEATHER  Adults only 18+  Everyone  Everyone 10+  \
8532          False    False            False      True         False   
324           False    False            False     False         False   
4541          False    False            False      True         False   
4636          False    False            False     False         False   
5940          False    False            False      True         False   

      Mature 17+   Teen  Unrated  
8532       False  False    False  
324         True  False    False  
4541       False  False    False  
4636        True  False    False  
5940       False  False    False  

[5 rows x 101 columns]

#### No Genres:

In [629]:
no_genres = clean_df.drop(['Genres', 'genre_1', 'genre_2'], axis=1)

In [634]:
no_genres = pd.concat([no_genres, content_rating], axis=1)
no_genres = no_genres.drop('Content Rating', axis=1)

In [635]:
no_genres.head()

App            Category  Rating  Reviews  \
8532  +Download 4 Instagram Twitter              SOCIAL     4.5    40467   
324      - Free Comics - Comic Apps              COMICS     3.5      115   
4541                             .R               TOOLS     4.5      259   
4636                         /u/app       COMMUNICATION     4.7      573   
5940                         058.ba  NEWS AND MAGAZINES     4.4       27   

       Size  Label  Paid  Price Last Updated  Adults only 18+  ...  \
8532  22000      1     0    0.0   2018-08-02            False  ...   
324   91000      0     0    0.0   2018-07-13            False  ...   
4541    203      0     0    0.0   2014-09-16            False  ...   
4636  53000      0     0    0.0   2018-07-03            False  ...   
5940  14000      0     0    0.0   2018-07-06            False  ...   

      Everyone 10+  Mature 17+   Teen  Unrated  Adults only 18+  Everyone  \
8532         False       False  False    False            False      True   
324          False        True  False    False            False     False   
4541         False       False  False    False            False      True   
4636         False        True  False    False            False     False   
5940         False       False  False    False            False      True   

      Everyone 10+  Mature 17+   Teen  Unrated  
8532         False       False  False    False  
324          False        True  False    False  
4541         False       False  False    False  
4636         False        True  False    False  
5940         False       False  False    False  

[5 rows x 21 columns]

#### One Hot + Label encoding:

In [648]:
ol = pd.concat([clean_df, genres_1, genres_2, content_rating], axis=1)

In [650]:
ol = ol.drop(['Genres', 'Content Rating', 'genre_1', 'genre_2'], axis=1)

In [651]:
ol.head()

App            Category  Rating  Reviews  \
8532  +Download 4 Instagram Twitter              SOCIAL     4.5    40467   
324      - Free Comics - Comic Apps              COMICS     3.5      115   
4541                             .R               TOOLS     4.5      259   
4636                         /u/app       COMMUNICATION     4.7      573   
5940                         058.ba  NEWS AND MAGAZINES     4.4       27   

       Size  Label  Paid  Price Last Updated  Action  ...  Creativity  \
8532  22000      1     0    0.0   2018-08-02   False  ...       False   
324   91000      0     0    0.0   2018-07-13   False  ...       False   
4541    203      0     0    0.0   2014-09-16   False  ...       False   
4636  53000      0     0    0.0   2018-07-03   False  ...       False   
5940  14000      0     0    0.0   2018-07-06   False  ...       False   

      Education  Music & Video  Pretend Play  Adults only 18+  Everyone  \
8532      False          False         False            False      True   
324       False          False         False            False     False   
4541      False          False         False            False      True   
4636      False          False         False            False     False   
5940      False          False         False            False      True   

      Everyone 10+  Mature 17+   Teen  Unrated  
8532         False       False  False    False  
324          False        True  False    False  
4541         False       False  False    False  
4636         False        True  False    False  
5940         False       False  False    False  

[5 rows x 69 columns]